pkg/client/connector/userd_trafficmgr/install_test.go

```diff
@@ -171,9 +176,137 @@ func (is *installSuite) Test_ensureTrafficManager_updateFromLegacy() {
+	ctx, err := client.SetDefaultConfig(ctx, is.T().TempDir())
+	require.NoError(err)
+
+	env, err := client.LoadEnv(ctx)
+	require.NoError(err)
+
+	cfgAndFlags, err := userd_k8s.NewConfig(map[string]string{"kubeconfig": is.kubeConfig, "namespace": is.managerNamespace}, env)
+	require.NoError(err)
+	kc, err := userd_k8s.NewCluster(ctx, cfgAndFlags, nil, userd_k8s.Callbacks{})
+	require.NoError(err)
+
+	version.Version = "v0.0.0-bogus"
+
+	restoreVersion := func() { version.Version = is.testVersion }
+
+	// We'll call this further down, but defer it to prevent polluting other tests if we don't leave this function gracefully
+	defer restoreVersion()
+
+	defer is.removeManager(is.managerNamespace)
+
+	ti, err := newTrafficManagerInstaller(kc)
+	require.NoError(err)
+
+	require.Error(ti.ensureManager(ctx, &env))
+	restoreVersion()
+
+	require.NoError(ti.ensureManager(ctx, &env))
+}
+
+func (is *installSuite) Test_ensureTrafficManager_toleratesLeftoverState() {
+	require := is.Require()
+	ctx := dlog.NewTestContext(is.T(), false)
+	ctx, err := client.SetDefaultConfig(ctx, is.T().TempDir())
+	require.NoError(err)
+
+	env, err := client.LoadEnv(ctx)
+	require.NoError(err)
+
+	cfgAndFlags, err := userd_k8s.NewConfig(map[string]string{"kubeconfig": is.kubeConfig, "namespace": is.managerNamespace}, env)
+	require.NoError(err)
+	kc, err := userd_k8s.NewCluster(ctx, cfgAndFlags, nil, userd_k8s.Callbacks{})
+	require.NoError(err)
+
+	ti, err := newTrafficManagerInstaller(kc)
+	require.NoError(err)
+
+	require.NoError(ti.ensureManager(ctx, &env))
+	defer is.removeManager(is.managerNamespace)
+	err = dexec.CommandContext(ctx, "../../../../tools/bin/helm", "--kubeconfig", is.kubeConfig, "--namespace", is.managerNamespace, "uninstall", "--keep-history", "traffic-manager").Run()
+	require.NoError(err)
+
+	require.NoError(ti.ensureManager(ctx, &env))
+	require.Eventually(func() bool {
+		deploy, err := ti.FindDeployment(ctx, is.managerNamespace, install.ManagerAppName)
+		if err != nil {
+			return false
+		}
+		return deploy.Status.ReadyReplicas == int32(1) && deploy.Status.Replicas == int32(1)
+	}, 10*time.Second, time.Second, "timeout waiting for deployment to update")
+}
+
+func (is *installSuite) Test_ensureTrafficManager_canUninstall() {
+	require := is.Require()
+	ctx := dlog.NewTestContext(is.T(), false)
+	ctx, err := client.SetDefaultConfig(ctx, is.T().TempDir())
+	require.NoError(err)
+
+	env, err := client.LoadEnv(ctx)
+	require.NoError(err)
+
+	cfgAndFlags, err := userd_k8s.NewConfig(map[string]string{"kubeconfig": is.kubeConfig, "namespace": is.managerNamespace}, env)
+	require.NoError(err)
+	kc, err := userd_k8s.NewCluster(ctx, cfgAndFlags, nil, userd_k8s.Callbacks{})
+	require.NoError(err)
+
+	ti, err := newTrafficManagerInstaller(kc)
+	require.NoError(err)
+
+	require.NoError(ti.ensureManager(ctx, &env))
+	require.NoError(ti.removeManagerAndAgents(ctx, false, []*manager.AgentInfo{}, &env))
+	require.NoError(ti.ensureManager(ctx, &env))
```

Comment:
Why is this sequence repeated?

Reply:
I noticed you had reported issues with uninstalling and then re-installing, so I wanted to make sure that was tested -- the final `removeManagerAndAgents` could probably be replaced by `is.removeManager`, since it's not _really_ under test, but I think we might as well exercise the uninstaller a second time